README.rst
.. image:: http://blog.ziade.org/boom.png

**Boom!** is a simple command line tool to send some load to a web app.

.. image:: https://img.shields.io/coveralls/tarekziade/boom.svg
    :target: https://coveralls.io/r/tarekziade/boom

.. image:: https://img.shields.io/travis/tarekziade/boom/master.svg
    :target: https://travis-ci.org/tarekziade/boom

.. image:: https://img.shields.io/pypi/v/boom.svg
    :target: https://pypi.python.org/pypi/boom

.. image:: https://img.shields.io/pypi/pyversions/boom.svg
    :target: https://pypi.python.org/pypi/boom/

.. image:: https://img.shields.io/pypi/dd/boom.svg
    :target: https://pypi.python.org/pypi/boom/

Boom! is a script you can use to quickly smoke-test your
web app deployment. If you need a more complex tool,
I'd suggest looking at `Funkload <http://funkload.nuxeo.org/>`_
or `Locust <https://github.com/locustio/locust>`_.

Boom! was specifically written to replace my Apache Bench usage,
to provide a few missing features and fix a few annoyances I had
with AB.

I have no special ambitions for this tool, and since I have not
found any tool like this in the Python-land, I wrote this one.

There are a lot of other tools out there, like Siege which
seems very popular.

However, Boom! is a good choice because it works on any platform
and is able to simulate thousands of users by using greenlets.

Installation
============

Boom! requires **Gevent** and **Requests**. If you are under Windows
I strongly recommend installing Gevent with the *xxx-win32-py2.7.exe*
installer you will find  at: https://github.com/surfly/gevent/downloads

Boom! should work with the latest versions.

If you are under Linux, installing the source version is usually a better
idea. You will need libev for Gevent.

Example under Ubuntu::

    $ sudo apt-get install libev libev-dev python-dev

Then::

    $ pip install boom


Basic usage
===========

Basic usage example: 100 queries with a maximum concurrency of
10 users::

    $ boom http://localhost:80 -c 10 -n 100
    Server Software: nginx/1.2.2
    Running 100 queries - concurrency: 10.
    Starting the load [===================================] Done

    -------- Results --------
    Successful calls        100
    Total time              0.3260 s
    Average                 0.0192 s
    Fastest                 0.0094 s
    Slowest                 0.0285 s
    Amplitude               0.0191 s
    RPS                     306
    BSI                     Pretty good

    -------- Legend --------
    RPS: Request Per Second
    BSI: Boom Speed Index


Boom! has more options::

    $ boom --help
    usage: boom [-h] [--version] [-m {GET,POST,DELETE,PUT,HEAD,OPTIONS}]
                [--content-type CONTENT_TYPE] [-D DATA] [-c CONCURRENCY] [-a AUTH]
                [--header HEADER] [--pre-hook PRE_HOOK] [--post-hook POST_HOOK]
                [--json-output] [-n REQUESTS | -d DURATION]
                [url]

    Simple HTTP Load runner.

    positional arguments:
      url                   URL to hit

    optional arguments:
      -h, --help            show this help message and exit
      --version             Displays version and exits.
      -m {GET,POST,DELETE,PUT,HEAD,OPTIONS}, --method {GET,POST,DELETE,PUT,HEAD,OPTIONS}
                            HTTP Method
      --content-type CONTENT_TYPE
                            Content-Type
      -D DATA, --data DATA  Data. Prefixed by "py:" to point a python callable.
      -c CONCURRENCY, --concurrency CONCURRENCY
                            Concurrency
      -a AUTH, --auth AUTH  Basic authentication user:password
      --header HEADER       Custom header. name:value
      --pre-hook PRE_HOOK   Python module path (eg: mymodule.pre_hook) to a
                            callable which will be executed before doing a request
                            for example: pre_hook(method, url, options). It must
                            return a tupple of parameters given in function
                            definition
      --post-hook POST_HOOK
                            Python module path (eg: mymodule.post_hook) to a
                            callable which will be executed after a request is
                            done for example: eg. post_hook(response). It must
                            return a given response parameter or raise an
                            `boom.boom.RequestException` for failed request.
      --json-output         Prints the results in JSON instead of the default
                            format
      -n REQUESTS, --requests REQUESTS
                            Number of requests
      -d DURATION, --duration DURATION
                            Duration in seconds



Design
======

Boom uses greenlets through Gevent to create *virtual users*, and uses Requests to do the
queries.

Using greenlets allows Boom to spawn large amounts of virtual users with very little
resources. It's not a problem to spawn 1000 users and hammer a web application with them.

If you are interested in this project, you are welcome to join the fun at
https://github.com/tarekziade/boom

Make sure to add yourself to the contributors list if your PR gets merged. And make sure it's in alphabetical order!


CHANGES.rst
History
=======

1.0 - 2016-09-05
----------------

- Update `.gitignore` for pycharm users
- Alphabetize contributors list for simplicity (read OCD)
- Replace BSI response of Hahahaha with sadface
- Add Travis CI
- Add Coveralls
- Update tox to test for 2.7 and 3.5 locally (travis ci handles the rest)


0.9 - 2016-08-28
----------------

- python 3 support
- removal of python 2.6 support
- fix unittest2 dep
- Added --validator option for validating request response data
- Change --hook to --pre-hook
- added new option --post-hook for validating request response data
- fixed the error handling of failed DNS resolution
- Replace urlparse.urlparse with urllib3's parse_url


0.8 - 2013-07-14
----------------

- Nicer progress bar
- Added the --json-output option
- Integrated Tox
- Make sure the DNS resolution works with gevent 0.x and 1.x
- Improved tests
- Removed the globals


0.7 - 2013-05-21
----------------

- fixed the TypeError on empty stats - #19
- catch the dns error and display it nicely
- added SSL support - #17
- added clean error display - #16

No notes on earlier releases.


boom/tests.py
import unittest2 as unittest
import subprocess
import sys
import shlex
import StringIO
import json

from gevent.pywsgi import WSGIServer
import requests
import gevent

from boom._boom import run as runboom, main, resolve, RunResults
from boom import _boom


class App(object):
    def __init__(self):
        self.numcalls = 0

    def handle(self, env, start_response):
        if env['PATH_INFO'] == '/':
            self.numcalls += 1
            start_response('200 OK', [('Content-Type', 'text/html')])
            return ["<b>hello world</b>"]
        elif env['PATH_INFO'] == '/calls':
            start_response('200 OK', [('Content-Type', 'text/plain')])
            return [str(self.numcalls)]
        elif env['PATH_INFO'] == '/redir':
            self.numcalls += 1
            start_response('302 Found', [('Location', '/redir')])
            return []
        elif env['PATH_INFO'] == '/reset':
            self.numcalls = 0
            start_response('200 OK', [('Content-Type', 'text/plain')])
            return ['numcalls set to zero']
        else:
            start_response('404 Not Found', [('Content-Type', 'text/plain')])
            return ['%s' % env['PATH_INFO']]


def run():
    app = App()
    WSGIServer(('0.0.0.0', 8089), app.handle).serve_forever()


CMD = shlex.split("%s -c 'from boom.tests import run; run()'" % sys.executable)
_SERVER = None


def _start():
    global _SERVER
    _SERVER = subprocess.Popen(CMD, stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE)


def _stop():
    global _SERVER
    if _SERVER is not None:
        _SERVER.terminate()
        _SERVER = None


def hook(method, url, options):
    options['files'] = {'file': open(__file__, 'rb')}
    return method, url, options


class TestBoom(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        _start()
        cls.server = 'http://0.0.0.0:8089'
        while True:
            try:
                requests.get(cls.server + '/')
                return
            except requests.ConnectionError:
                gevent.sleep(.1)

    @classmethod
    def tearDownClass(cls):
        _stop()

    def setUp(self):
        self.get('/reset')

    def get(self, path):
        return requests.get(self.server + path)

    def test_basic_run(self):
        runboom(self.server, num=10, concurrency=1)
        res = self.get('/calls').content
        self.assertEqual(int(res), 10)

    def test_hook(self):
        runboom(self.server, method='POST', num=10, concurrency=1,
                hook='boom.tests.hook')
        res = self.get('/calls').content
        self.assertEqual(int(res), 10)

    def test_connection_error(self):
        run_results = runboom('http://localhost:9999', num=10, concurrency=1)
        self.assertEqual(len(run_results.errors), 10)
        for error in run_results.errors:
            self.assertIsInstance(error, requests.ConnectionError)

    def test_too_many_redirects(self):
        run_results = runboom(self.server + '/redir', num=2, concurrency=1)
        res = self.get('/calls').content
        self.assertEqual(int(res), 62)
        for error in run_results.errors:
            self.assertIsInstance(error, requests.TooManyRedirects)

    def _run(self, *args):
        sys.argv[:] = [sys.executable] + list(args)
        old_stdout = sys.stdout
        old_stderr = sys.stderr
        exit_code = 0
        sys.stdout = StringIO.StringIO()
        sys.stderr = StringIO.StringIO()
        try:
            main()
        except (Exception, SystemExit), e:
            if isinstance(e, SystemExit):
                exit_code = e.code
            else:
                exit_code = 1
        finally:
            sys.stdout.seek(0)
            stdout = sys.stdout.read()
            sys.stderr.seek(0)
            stderr = sys.stdout.read()
            sys.stdout = old_stdout
            sys.stderr = old_stderr

        return exit_code, stdout, stderr

    def test_dns_resolve(self):
        code, stdout, stderr = self._run('http://that.impossiblename')
        self.assertEqual(code, 1)
        self.assertTrue('DNS resolution failed for' in stdout, stdout)

    def test_resolve(self):
        test_url = 'http://localhost:9999'
        url, original, resolved = resolve(test_url)
        self.assertEqual(url, 'http://127.0.0.1:9999')
        self.assertEqual(original, 'localhost')
        self.assertEqual(resolved, '127.0.0.1')

    def test_ssl_resolve(self):
        test_url = 'https://localhost:9999'
        url, original, resolved = resolve(test_url)
        self.assertEqual(url, 'https://localhost:9999')
        self.assertEqual(original, 'localhost')
        self.assertEqual(resolved, 'localhost')

    def test_resolve_no_scheme(self):
        test_url = 'http://localhost'
        url, original, resolved = resolve(test_url)
        self.assertEqual(url, 'http://127.0.0.1:80')
        self.assertEqual(original, 'localhost')
        self.assertEqual(resolved, '127.0.0.1')

    def test_resolve_no_scheme_ssl(self):
        test_url = 'https://localhost'
        url, original, resolved = resolve(test_url)
        self.assertEqual(url, 'https://localhost:443')
        self.assertEqual(original, 'localhost')
        self.assertEqual(resolved, 'localhost')

    def test_json_output(self):
        results = RunResults()
        results.status_code_counter['200'].extend([0, 0.1, 0.2])
        results.total_time = 9

        old_stdout = sys.stdout
        old_stderr = sys.stderr

        try:
            sys.stdout = StringIO.StringIO()
            sys.stderr = StringIO.StringIO()
            _boom.print_json(results)

            sys.stdout.seek(0)
            output = sys.stdout.read()
        finally:
            sys.stdout = old_stdout
            sys.stderr = old_stderr

        actual = json.loads(output)
        self.assertEqual(3, actual['count'])
        self.assertAlmostEqual(0.333, actual['rps'], delta=0.1)
        self.assertEqual(0, actual['min'])
        self.assertEqual(0.2, actual['max'])
        self.assertAlmostEqual(0.1, actual['avg'], delta=0.1)
        self.assertEqual(0.2, actual['amp'])


if __name__ == '__main__':
    run()


boom/tests/__init__.py
# -*- coding: utf-8 -*-
import unittest


def load_tests(loader, tests, pattern):
    """ Discover and load all unit tests in all files named ``test_*.py`` """
    suite = unittest.TestSuite()
    for all_test_suite in unittest.defaultTestLoader.discover('.', pattern='test_*.py'):
        for test_suite in all_test_suite:
            suite.addTests(test_suite)
    return suite

if __name__ == '__main__':
    unittest.main()


boom/tests/test_boom.py
import unittest2 as unittest
import subprocess
import sys
import shlex
try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO
import json

from gevent.pywsgi import WSGIServer
import requests
import gevent

from boom.boom import (run as runboom, main,
                       resolve, RunResults, RequestException)
from boom import boom


if sys.version_info[0] < 3:
    PY3 = False
else:
    PY3 = True


class App(object):

    def __init__(self):
        self.numcalls = 0

    def handle(self, env, start_response):
        if env['PATH_INFO'] == '/':
            self.numcalls += 1
            start_response('200 OK', [('Content-Type', 'text/html')])
            if PY3:
                return ["<b>hello world</b>".encode('latin-1')]
            else:
                return ["<b>hello world</b>"]
        elif env['PATH_INFO'] == '/calls':
            start_response('200 OK', [('Content-Type', 'text/plain')])
            if PY3:
                return [str(self.numcalls).encode('latin-1')]
            else:
                return [str(self.numcalls)]
        elif env['PATH_INFO'] == '/redir':
            self.numcalls += 1
            start_response('302 Found', [('Location', '/redir')])
            return []
        elif env['PATH_INFO'] == '/reset':
            self.numcalls = 0
            start_response('200 OK', [('Content-Type', 'text/plain')])
            if PY3:
                return ['numcalls set to zero'.encode('latin-1')]
            else:
                return ['numcalls set to zero']
        else:
            start_response('404 Not Found',
                           [('Content-Type', 'text/plain')])
            if PY3:
                return [env['PATH_INFO'].encode('latin-1')]
            else:
                return [env['PATH_INFO']]


def run():
    app = App()
    WSGIServer(('0.0.0.0', 8089), app.handle).serve_forever()


_CMD = "%s -c 'from boom.tests.test_boom import run; run()'"
CMD = shlex.split(_CMD % sys.executable)
_SERVER = None


def _start():
    global _SERVER
    _SERVER = subprocess.Popen(CMD, stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE)


def _stop():
    global _SERVER
    if _SERVER is not None:
        _SERVER.terminate()
        _SERVER = None


def pre_hook(method, url, options):
    options['files'] = {'file': open(__file__, 'rb')}
    return method, url, options


def post_hook(response):
    return response


def post_hook_fails(data):
    if 'pattern' not in data:
        raise RequestException('missing pattern')
    return data


class TestBoom(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        _start()
        cls.server = 'http://0.0.0.0:8089'
        while True:
            try:
                requests.get(cls.server + '/')
                return
            except requests.ConnectionError:
                gevent.sleep(.1)

    @classmethod
    def tearDownClass(cls):
        _stop()

    def setUp(self):
        self.get('/reset')

    def get(self, path):
        return requests.get(self.server + path)

    def test_basic_run(self):
        runboom(self.server, num=10, concurrency=1, quiet=True)
        res = self.get('/calls').content
        self.assertEqual(int(res), 10)

    def test_pre_hook(self):
        runboom(self.server, method='POST', num=10, concurrency=1,
                pre_hook='boom.tests.test_boom.pre_hook', quiet=True)
        res = self.get('/calls').content
        self.assertEqual(int(res), 10)

    def test_post_hook(self):
        run_results = runboom(
            self.server, method='GET', num=10, concurrency=1,
            post_hook='boom.tests.test_boom.post_hook', quiet=True)
        res = self.get('/calls').content
        self.assertEqual(run_results.errors, [])
        self.assertEqual(int(res), 10)

    def test_post_hook_fails(self):
        run_results = runboom(
            self.server, method='GET', num=10, concurrency=1,
            post_hook='boom.tests.test_boom.post_hook_fails', quiet=True)
        res = self.get('/calls').content
        self.assertEqual(len(run_results.errors), 10)

        for err in run_results.errors:
            self.assertEqual(True, isinstance(err, RequestException))
            self.assertEqual(err.__str__(), 'missing pattern')

        self.assertEqual(int(res), 10)

    def test_connection_error(self):
        run_results = runboom(
            'http://localhost:9999', num=10, concurrency=1,
            quiet=True)
        self.assertEqual(len(run_results.errors), 10)
        for error in run_results.errors:
            self.assertIsInstance(error, requests.ConnectionError)

    def test_too_many_redirects(self):
        run_results = runboom(
            self.server + '/redir', num=2, concurrency=1,
            quiet=True)
        res = self.get('/calls').content
        self.assertEqual(int(res), 62)
        for error in run_results.errors:
            self.assertIsInstance(error, requests.TooManyRedirects)

    def _run(self, *args):
        sys.argv[:] = [sys.executable] + list(args)
        old_stdout = sys.stdout
        old_stderr = sys.stderr
        exit_code = 0
        sys.stdout = StringIO()
        sys.stderr = StringIO()
        try:
            main()
        except (Exception, SystemExit) as e:
            if isinstance(e, SystemExit):
                exit_code = e.code
            else:
                exit_code = 1
        finally:
            sys.stdout.seek(0)
            stdout = sys.stdout.read()
            sys.stderr.seek(0)
            stderr = sys.stdout.read()
            sys.stdout = old_stdout
            sys.stderr = old_stderr

        return exit_code, stdout, stderr

    def test_dns_resolve(self):
        code, stdout, stderr = self._run('http://that.impossiblename')
        self.assertEqual(code, 1)
        self.assertTrue('DNS resolution failed for' in stdout, stdout)

    def test_resolve(self):
        test_url = 'http://localhost:9999'
        url, original, resolved = resolve(test_url)
        self.assertEqual(url, 'http://127.0.0.1:9999')
        self.assertEqual(original, 'localhost:9999')
        self.assertEqual(resolved, '127.0.0.1:9999')

    def test_ssl_resolve(self):
        test_url = 'https://localhost:9999'
        url, original, resolved = resolve(test_url)
        self.assertEqual(url, 'https://localhost:9999')
        self.assertEqual(original, 'localhost:9999')
        self.assertEqual(resolved, 'localhost:9999')

    def test_resolve_no_scheme(self):
        test_url = 'http://localhost'
        url, original, resolved = resolve(test_url)
        self.assertEqual(url, 'http://127.0.0.1:80')
        self.assertEqual(original, 'localhost')
        self.assertEqual(resolved, '127.0.0.1')

    def test_resolve_no_scheme_ssl(self):
        test_url = 'https://localhost'
        url, original, resolved = resolve(test_url)
        self.assertEqual(url, 'https://localhost:443')
        self.assertEqual(original, 'localhost')
        self.assertEqual(resolved, 'localhost')

    def test_json_output(self):
        results = RunResults()
        results.status_code_counter['200'].extend([0, 0.1, 0.2])
        results.total_time = 9

        old_stdout = sys.stdout
        old_stderr = sys.stderr

        try:
            sys.stdout = StringIO()
            sys.stderr = StringIO()
            boom.print_json(results)
            sys.stdout.seek(0)
            output = sys.stdout.read()
        finally:
            sys.stdout = old_stdout
            sys.stderr = old_stderr

        actual = json.loads(output)
        self.assertEqual(3, actual['count'])
        self.assertAlmostEqual(0.333, actual['rps'], delta=0.1)
        self.assertEqual(0, actual['min'])
        self.assertEqual(0.2, actual['max'])
        self.assertAlmostEqual(0.1, actual['avg'], delta=0.1)
        self.assertEqual(0.2, actual['amp'])


if __name__ == '__main__':
    unittest.main()


boom/boom.py
from __future__ import absolute_import
import argparse
import gevent
import logging
import requests
import sys
import time

try:
    import urlparse
except ImportError:
    from urllib import parse as urlparse

import math
from collections import defaultdict, namedtuple
from copy import copy
from gevent import monkey
from gevent.pool import Pool
from requests import RequestException
from requests.packages.urllib3.util import parse_url
from socket import gethostbyname, gaierror

from boom import __version__
from boom.util import resolve_name
from boom.pgbar import AnimatedProgressBar


monkey.patch_all()
logger = logging.getLogger('boom')
_VERBS = ('GET', 'POST', 'DELETE', 'PUT', 'HEAD', 'OPTIONS')
_DATA_VERBS = ('POST', 'PUT')


class RunResults(object):

    """Encapsulates the results of a single Boom run.

    Contains a dictionary of status codes to lists of request durations,
    a list of exception instances raised during the run, the total time
    of the run and an animated progress bar.
    """

    def __init__(self, num=1, quiet=False):
        self.status_code_counter = defaultdict(list)
        self.errors = []
        self.total_time = None
        if num is not None:
            self._progress_bar = AnimatedProgressBar(
                end=num,
                width=65)
        else:
            self._progress_bar = None
        self.quiet = quiet

    def incr(self):
        if self.quiet:
            return
        if self._progress_bar is not None:
            self._progress_bar + 1
            self._progress_bar.show_progress()
        else:
            sys.stdout.write('.')
            sys.stdout.flush()


RunStats = namedtuple(
    'RunStats', ['count', 'total_time', 'rps', 'avg', 'min',
                 'max', 'amp', 'stdev'])


def calc_stats(results):
    """Calculate stats (min, max, avg) from the given RunResults.

       The statistics are returned as a RunStats object.
    """
    all_res = []
    count = 0
    for values in results.status_code_counter.values():
        all_res += values
        count += len(values)

    cum_time = sum(all_res)

    if cum_time == 0 or len(all_res) == 0:
        rps = avg = min_ = max_ = amp = stdev = 0
    else:
        if results.total_time == 0:
            rps = 0
        else:
            rps = len(all_res) / float(results.total_time)
        avg = sum(all_res) / len(all_res)
        max_ = max(all_res)
        min_ = min(all_res)
        amp = max(all_res) - min(all_res)
        stdev = math.sqrt(sum((x-avg)**2 for x in all_res) / count)

    return (
        RunStats(count, results.total_time, rps, avg, min_, max_, amp, stdev)
    )


def print_stats(results):
    stats = calc_stats(results)
    rps = stats.rps

    print('')
    print('-------- Results --------')

    print('Successful calls\t\t%r' % stats.count)
    print('Total time        \t\t%.4f s  ' % stats.total_time)
    print('Average           \t\t%.4f s  ' % stats.avg)
    print('Fastest           \t\t%.4f s  ' % stats.min)
    print('Slowest           \t\t%.4f s  ' % stats.max)
    print('Amplitude         \t\t%.4f s  ' % stats.amp)
    print('Standard deviation\t\t%.6f' % stats.stdev)
    print('RPS               \t\t%d' % rps)
    if rps > 500:
        print('BSI              \t\tWoooooo Fast')
    elif rps > 100:
        print('BSI              \t\tPretty good')
    elif rps > 50:
        print('BSI              \t\tMeh')
    else:
        print('BSI              \t\t:(')
    print('')
    print('-------- Status codes --------')
    for code, items in results.status_code_counter.items():
        print('Code %d          \t\t%d times.' % (code, len(items)))
    print('')
    print('-------- Legend --------')
    print('RPS: Request Per Second')
    print('BSI: Boom Speed Index')


def print_server_info(url, method, headers=None):
    res = requests.head(url)
    print(
        'Server Software: %s' %
        res.headers.get('server', 'Unknown'))
    print('Running %s %s' % (method, url))

    if headers:
        for k, v in headers.items():
            print('\t%s: %s' % (k, v))


def print_errors(errors):
    if len(errors) == 0:
        return
    print('')
    print('-------- Errors --------')
    for error in errors:
        print(error)


def print_json(results):
    """Prints a JSON representation of the results to stdout."""
    import json
    stats = calc_stats(results)
    print(json.dumps(stats._asdict()))


def onecall(method, url, results, **options):
    """Performs a single HTTP call and puts the result into the
       status_code_counter.

    RequestExceptions are caught and put into the errors set.
    """
    start = time.time()

    if 'data' in options and callable(options['data']):
        options = copy(options)
        options['data'] = options['data'](method, url, options)

    if 'pre_hook' in options:
        method, url, options = options[
            'pre_hook'](method, url, options)
        del options['pre_hook']

    if 'post_hook' in options:
        post_hook = options['post_hook']
        del options['post_hook']
    else:
        def post_hook(res):
            return res

    try:
        res = post_hook(method(url, **options))
    except RequestException as exc:
        results.errors.append(exc)
    else:
        duration = time.time() - start
        results.status_code_counter[res.status_code].append(duration)
    finally:
        results.incr()


def run(
    url, num=1, duration=None, method='GET', data=None, ct='text/plain',
        auth=None, concurrency=1, headers=None, pre_hook=None, post_hook=None,
        quiet=False):

    if headers is None:
        headers = {}

    if 'content-type' not in headers:
        headers['Content-Type'] = ct

    if data is not None and data.startswith('py:'):
        callable = data[len('py:'):]
        data = resolve_name(callable)

    method = getattr(requests, method.lower())
    options = {'headers': headers}

    if pre_hook is not None:
        options['pre_hook'] = resolve_name(pre_hook)

    if post_hook is not None:
        options['post_hook'] = resolve_name(post_hook)

    if data is not None:
        options['data'] = data

    if auth is not None:
        options['auth'] = tuple(auth.split(':', 1))

    pool = Pool(concurrency)
    start = time.time()
    jobs = None
    res = RunResults(num, quiet)

    try:
        if num is not None:
            jobs = [pool.spawn(onecall, method, url, res, **options)
                    for i in range(num)]
            pool.join()
        else:
            with gevent.Timeout(duration, False):
                jobs = []
                while True:
                    jobs.append(pool.spawn(onecall, method, url, res,
                                           **options))
                pool.join()
    except KeyboardInterrupt:
        # In case of a keyboard interrupt, just return whatever already got
        # put into the result object.
        pass
    finally:
        res.total_time = time.time() - start

    return res


def resolve(url):
    parts = parse_url(url)

    if not parts.port and parts.scheme == 'https':
        port = 443
    elif not parts.port and parts.scheme == 'http':
        port = 80
    else:
        port = parts.port

    original = parts.host
    resolved = gethostbyname(parts.host)

    # Don't use a resolved hostname for SSL requests otherwise the
    # certificate will not match the IP address (resolved)
    host = resolved if parts.scheme != 'https' else parts.host
    netloc = '%s:%d' % (host, port) if port else host

    if port not in (443, 80):
        host += ':%d' % port
        original += ':%d' % port

    return (urlparse.urlunparse((parts.scheme, netloc, parts.path or '',
                                 '', parts.query or '',
                                 parts.fragment or '')),
            original, host)


def load(url, requests, concurrency, duration, method, data, ct, auth,
         headers=None, pre_hook=None, post_hook=None, quiet=False):
    if not quiet:
        print_server_info(url, method, headers=headers)

        if requests is not None:
            print('Running %d queries - concurrency %d' % (requests,
                                                           concurrency))
        else:
            print('Running for %d seconds - concurrency %d.' %
                  (duration, concurrency))

        sys.stdout.write('Starting the load')
    try:
        return run(url, requests, duration, method,
                   data, ct, auth, concurrency, headers,
                   pre_hook, post_hook, quiet=quiet)
    finally:
        if not quiet:
            print(' Done')


def main():
    parser = argparse.ArgumentParser(
        description='Simple HTTP Load runner.')

    parser.add_argument(
        '--version', action='store_true', default=False,
        help='Displays version and exits.')

    parser.add_argument('-m', '--method', help='HTTP Method',
                        type=str, default='GET', choices=_VERBS)

    parser.add_argument('--content-type', help='Content-Type',
                        type=str, default='text/plain')

    parser.add_argument('-D', '--data',
                        help=('Data. Prefixed by "py:" to point '
                              'a python callable.'),
                        type=str)

    parser.add_argument('-c', '--concurrency', help='Concurrency',
                        type=int, default=1)

    parser.add_argument('-a', '--auth',
                        help='Basic authentication user:password', type=str)

    parser.add_argument('--header', help='Custom header. name:value',
                        type=str, action='append')

    parser.add_argument('--pre-hook',
                        help=("Python module path (eg: mymodule.pre_hook) "
                              "to a callable which will be executed before "
                              "doing a request for example: "
                              "pre_hook(method, url, options). "
                              "It must return a tupple of parameters given in "
                              "function definition"),
                        type=str)

    parser.add_argument('--post-hook',
                        help=("Python module path (eg: mymodule.post_hook) "
                              "to a callable which will be executed after "
                              "a request is done for example: "
                              "eg. post_hook(response). "
                              "It must return a given response parameter or "
                              "raise an `boom.boom.RequestException` for "
                              "failed request."),
                        type=str)

    parser.add_argument('--json-output',
                        help='Prints the results in JSON instead of the '
                             'default format',
                        action='store_true')

    parser.add_argument('-q', '--quiet', help="Don't display progress bar",
                        action='store_true')

    group = parser.add_mutually_exclusive_group()

    group.add_argument('-n', '--requests', help='Number of requests',
                       type=int)

    group.add_argument('-d', '--duration', help='Duration in seconds',
                       type=int)

    parser.add_argument('url', help='URL to hit', nargs='?')
    args = parser.parse_args()

    if args.version:
        print(__version__)
        sys.exit(0)

    if args.url is None:
        print('You need to provide an URL.')
        parser.print_usage()
        sys.exit(0)

    if args.data is not None and args.method not in _DATA_VERBS:
        print("You can't provide data with %r" % args.method)
        parser.print_usage()
        sys.exit(0)

    if args.requests is None and args.duration is None:
        args.requests = 1

    try:
        url, original, resolved = resolve(args.url)
    except gaierror as e:
        print_errors(("DNS resolution failed for %s (%s)" %
                      (args.url, str(e)),))
        sys.exit(1)

    def _split(header):
        header = header.split(':')

        if len(header) != 2:
            print("A header must be of the form name:value")
            parser.print_usage()
            sys.exit(0)

        return header

    if args.header is None:
        headers = {}
    else:
        headers = dict([_split(header) for header in args.header])

    if original != resolved and 'Host' not in headers:
        headers['Host'] = original

    try:
        res = load(
            url, args.requests, args.concurrency, args.duration,
            args.method, args.data, args.content_type, args.auth,
            headers=headers, pre_hook=args.pre_hook,
            post_hook=args.post_hook, quiet=(args.json_output or args.quiet))
    except RequestException as e:
        print_errors((e, ))
        sys.exit(1)

    if not args.json_output:
        print_errors(res.errors)
        print_stats(res)
    else:
        print_json(res)

    logger.info('Bye!')


if __name__ == '__main__':
    main()
